src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs

```diff
@@ -28,16 +28,67 @@ internal SwitchExpressionBinder(SwitchExpressionSyntax switchExpressionSyntax, B
-            switchCases = AddConversionsToCases(switchCases, resultType, diagnostics);
-            // PROTOTYPE(patterns2): check for subsumption, completeness, etc.
-            //var hasErrors = CheckSwitchErrors(node, boundSwitchGoverningExpression, switchCases, decisionDag, diagnostics);
-            return new BoundSwitchExpression(node, boundSwitchGoverningExpression, switchCases, resultType, hasErrors);
+            var boundInputExpression = InputExpression;
+            diagnostics.AddRange(InputExpressionDiagnostics);
+            ImmutableArray<BoundSwitchExpressionArm> switchArms = BindSwitchExpressionArms(node, originalBinder, diagnostics);
+            TypeSymbol resultType = InferResultType(switchArms, diagnostics);
+            switchArms = AddConversionsToArms(switchArms, resultType, diagnostics);
+            bool hasErrors = CheckSwitchExpressionExhaustive(node, boundInputExpression, switchArms, out BoundDecisionDag decisionDag, out LabelSymbol defaultLabel, diagnostics);
+            if (boundInputExpression.ConstantValue != null)
+            {
+                // When the input is constant, we use that to reshape the decision dag that is returned
+                // so that flow analysis will see that some of the cases may be unreachable.
+                decisionDag = decisionDag.SimplifyDecisionDagForConstantInput(boundInputExpression, Conversions, diagnostics);
+            }
+
+            return new BoundSwitchExpression(node, boundInputExpression, switchArms, decisionDag, defaultLabel, resultType, hasErrors);
+        }
+
+        /// <summary>
+        /// Build the decision dag, giving an error if some cases are subsumed and an error if the switch expression is not exhaustive.
+        /// Returns true if there were errors.
+        /// </summary>
+        /// <param name="node"></param>
+        /// <param name="boundInputExpression"></param>
+        /// <param name="switchArms"></param>
+        /// <param name="decisionDag"></param>
+        /// <param name="diagnostics"></param>
+        /// <returns>true if the switch expression is not exhaustive</returns>
+        private bool CheckSwitchExpressionExhaustive(
+            SwitchExpressionSyntax node,
+            BoundExpression boundInputExpression,
+            ImmutableArray<BoundSwitchExpressionArm> switchArms,
+            out BoundDecisionDag decisionDag,
+            out LabelSymbol defaultLabel,
+            DiagnosticBag diagnostics)
+        {
+            defaultLabel = new GeneratedLabelSymbol("default");
+            decisionDag = DecisionDagBuilder.CreateDecisionDag(this.Compilation, node, boundInputExpression, switchArms, defaultLabel, diagnostics);
+            HashSet<LabelSymbol> reachableLabels = decisionDag.ReachableLabels;
+            foreach (BoundSwitchExpressionArm arm in switchArms)
+            {
+                if (!reachableLabels.Contains(arm.Label))
```

Comment:
For the switch statement it looks like we only report subsumption if the label doesn't have errors. Why not here? Similarly, we aren't marking subsumed labels as errors, as in the switch statement.

Is it possible for the switch expression and the switch statement to share this code? It looks very similar.

Reply:
We need that handling for the switch statement to duplicate the diagnostics from previous versions of the compiler. We don't have those constraints here, though once we have switch expressions working and can start testing them, I expect we will want to suppress the subsumption error for erroneous cases.
Recorded in [https://github.com/dotnet/roslyn/issues/25663](https://github.com/dotnet/roslyn/issues/25663)

---
In reply to: [176224663](https://github.com/dotnet/roslyn/pull/24059#discussion_r176224663) [](ancestors = 176224663)